alchemiscale_fah/protocols/protocolunit.py

```diff
@@ -0,0 +1,284 @@
+            ctx.index.set_transformation(
+                ctx.transformation_sk.gufe_key, project_id, run_id
+            )
+
+        # if we got PROJECT and RUN IDs, but no CLONE ID, it means this Task
+        # has never been seen before on this work server, but the
+        # Transformation has; we use the existing PROJECT and RUN but create a
+        # new CLONE
+        if clone_id is None:
+            # create core file from settings
+            core_file_content = self.generate_core_settings_file(settings)
+
+            # get next available CLONE id
+            clone_id = ctx.index.get_run_clone_next(project_id, run_id)
+
+            # create CLONE for this Task
+            ctx.fah_client.create_clone_file_from_bytes(
+                project_id,
+                run_id,
+                clone_id,
+                str(ctx.task_sk).encode("utf-8"),
+                "alchemiscale-task.txt",
+            )
+
+            # TODO: add encryption of files here if enabled as a setting on the
+            # service use configured public key
+            if ctx.encryption_public_key:
+                ...
+
+            else:
+                ctx.fah_client.create_clone_file_from_bytes(
+                    project_id,
+                    run_id,
+                    clone_id,
+                    str(core_file_content).encode("utf-8"),
+                    "core.xml",
+                )
+
+                for filepath in (system_file, state_file, integrator_file):
+                    ctx.fah_client.create_clone_file(
+                        project_id, run_id, clone_id, filepath, filepath.name
+                    )
+
+            ctx.index.set_task(ctx.task_sk, project_id, run_id, clone_id)
+            ctx.fah_client.create_clone(project_id, run_id, clone_id)
+
+        while True:
+            # check for and await sleep results from work server
+            jobdata = ctx.fah_client.get_clone(project_id, run_id, clone_id)
+
+            if JobStateEnum[jobdata.state] is JobStateEnum.FINISHED:
+                break
+            elif JobStateEnum[jobdata.state] is JobStateEnum.FAILED:
+                raise FahExecutionException(
+                    "Consecutive failed or faulty WUs exceeded the "
+                    f"maximum for RUN {run_id} in PROJECT {project_id}"
+                )
+
+            else:
+                await asyncio.sleep(ctx.fah_poll_sleep)
+
+        # read in results from `globals.csv`
+        globals_csv = ctx.fah_client.get_gen_output_file_to_bytes(
+            project_id, run_id, clone_id, 0, settings.fah_settings.globalVarFilename
+        )
+
+        outputs = self.postprocess_globals(globals_csv, ctx)
+
+        # read in science log
+        science_log = ctx.fah_client.get_gen_output_file_to_bytes(
+            project_id, run_id, clone_id, 0, "science.log"
+        )
+
+        science_log_path = ctx.shared / "science.log"
+        with open(ctx.shared / "science.log", "wb") as f:
```

Comment:
Is this async writing all the `science.logs` to one spot? If so are the writes atomic or will they end up jumbled?

Reply:
No, the science logs are written to the `shared` dir for each `ProtocolUnit`, which is distinct.